cgen error on `return match op { '1' { fn() string { return '1 passed' }} ... }`

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Ubuntu 20.04.3 LTS
Processor: 2 cpus, 64bit, little endian, Intel(R) Core(TM) i3-3225 CPU @ 3.30GHz
CC version: cc (Ubuntu 10.3.0-1ubuntu1~20.04) 10.3.0

getwd: /v/cleanv
vmodules: /home/delian/.vmodules
vroot: /v/cleanv
vexe: /v/cleanv/v
vexe mtime: 2022-04-24 09:35:47
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 a2db44b.9355048

Git version: git version 2.25.1
Git vroot status: weekly.2022.16-47-g9355048b
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 eee03626

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg bug.v`
```v
fn foofun(op rune) fn() string{
	return match op {
		`1` { fn() string { return '1 passed' }}
		`2` { fn() string { return '2 passed'}}
		else { fn() string { return 'Nor 1 or 2 passed'}}
	}
}
 
fn main() {
	a := foofun(`1`)
	println(a())
}
 

```

**What did you expect to see?**

compiled program 


**What did you see instead?**
```
==================
/tmp/v_1000/bug.5127998886254470210.tmp.c:1941: error: ')' expected (got "_t1")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -379,7 +379,10 @@ fn (mut g Gen) gen_fn_decl(node &ast.FnDecl, skip bool) {
 	defer {
 		g.tmp_count = ctmp
 	}
+	prev_inside_ternary := g.inside_ternary
+	g.inside_ternary = 0
 	g.stmts(node.stmts)
+	g.inside_ternary = prev_inside_ternary
 	if node.is_noreturn {
 		g.writeln('\twhile(1);')
 	}

--- a/vlib/v/tests/anon_fn_decl_inside_ternary_test.v
+++ b/vlib/v/tests/anon_fn_decl_inside_ternary_test.v
@@ -1,0 +1,25 @@
+fn foofun(op rune) fn () string {
+	return match op {
+		`1` {
+			fn () string {
+				return '1 passed'
+			}
+		}
+		`2` {
+			fn () string {
+				return '2 passed'
+			}
+		}
+		else {
+			fn () string {
+				return 'Nor 1 or 2 passed'
+			}
+		}
+	}
+}
+
+fn test_anon_fn_decl_inside_ternary() {
+	a := foofun(`1`)
+	println(a())
+	assert a() == '1 passed'
+}

